File: TopInterviewMedium/Arrays/3Sum.cs
/*
https://leetcode.com/explore/interview/card/top-interview-questions-medium/103/array-and-strings/776/
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] 
such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.
- [0,0,0] -> [0,0,0]
- [1,0,0,0,-1] -> [1,0,-1],[0,0,0]
- [-1,0,1,2,-1,-4]
	[-1,0,1] | []
- [-1,0,1,2,-1,-4,-2,-3,3,0,4]
*/
public class Solution{
	public IList<IList<int>> ThreeSum(int[] nums) {
		var result  = new List<IList<int>>();
		var set = new HashSet<string>();
		var duplicateSet = new HashSet<int>();
        	for(int i=0;i<nums.Length;i++){
        		if(!duplicateSet.Add(nums[i]))
        			continue;
        		int sum = nums[i];
        		sum = (-1)*sum;
        		var sumElementsList = sum2(sum,nums,i);
				if(sumElementsList.Count == 0)
					continue;
                foreach (var sumValues in sumElementsList)
                {
                	var triplet = new List<int>();
                	bool numAdded = false;
                	for(int j = 0 ;j<2;j++)
					{
						if(nums[i]<sumValues[j] && !numAdded)
						{
							triplet.Add(nums[i]);
							numAdded = true;
						}
						triplet.Add(sumValues[j]);
					}
					if(triplet.Count<3)
						triplet.Add(nums[i]);
					
					string key = string.Join(",",triplet);
					if(!set.Contains(key))
					{
						result.Add(triplet);
						set.Add(key);
					}	
                }
        	}
			//Console.WriteLine(string.Join(";",set));
        	return result;
    }
	
	List<int[]> sum2(int sum, int[] nums, int ignoreIndex)
	{
		var result  = new List<int[]>();
		var set = new HashSet<int>();
		for(int i = 0 ; i<nums.Length; i++)
		{
			if(i==ignoreIndex)
				continue;
			int target = sum - nums[i];
			if(set.Contains(target))
			{
				if(target<nums[i])
					result.Add(new int[]{target,nums[i]});
				else
					result.Add(new int[]{nums[i],target});
			}
			set.Add(nums[i]);
		}
		return result;
	}
}

//2nd approach
// [-1,0,1,2,-1,-4]
// [-4,-1,-1,0,1,2]
public class Solution{
	public IList<IList<int>> ThreeSum(int[] nums) {
		IList<IList<int>> result = new List<IList<int>>();
		if(nums.Length <3)
			return result;

		Array.Sort(nums);
        //Console.WriteLine(string.Join(";",nums));
		int i = 0;
        while(i< nums.Length-2)
		{
			int currentValue = nums[i];
			int target = currentValue*-1;
			int left = i+1;
			int right = nums.Length-1;
            //Console.WriteLine($" start : {i}, {left}, {right}");
			while(left<right)
			{
				int value = nums[left] + nums[right];
                //Console.WriteLine($" value : {value}, {target}");
				if(target<value)
				{
					right--;
				}
				else if(target>value)
				{
					left++;
				}
				else
				{
                    
					int leftValue = nums[left];
					int rightValue = nums[right];
                    //Console.WriteLine($"{currentValue},{leftValue},{rightValue}");
					result.Add(new List<int>{currentValue,leftValue,rightValue});
					while((left<right) && nums[left] == leftValue)
					{
						left++;
					}		
					while((right>left) && nums[right] == rightValue)
					{
						right--;
					}

				}
			}
			while((i<nums.Length-2) && nums[i] == currentValue)
			{
				i++;
			}
		}
		return result;
    }
}

File: Grind75/3sum.py
"""
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

Notice that the solution set must not contain duplicate triplets.


[1, 2, 2, 3, -4] -> [[1,3,-4], [2,2,-4]]

indices_by_val = [(1,0), (2,[1,2]), (3,3), (-4,4)]

result : []
"""

class Solution:
	def threeSum(self, nums: List[int]) -> List[List[int]]:
		indices_by_val = dict()
		used_values = set() 
		for idx in range(0, len(nums)):
			if nums[idx] not in indices_by_val:
				indices_by_val[nums[idx]] = idx
		
		result = []
		for i in range(0, len(nums)-1):
			for j in range(i+1, len(nums)):
				if -(nums[i] + nums[j]) in indices_by_val and (nums[i], nums[j]) not in used_values and (nums[j], nums[i]) not in used_values:
					k =  indices_by_val[-(nums[i] + nums[j])]
					if i != k and j!=k:
						result.append([nums[i],nums[j],nums[k]])
						used_values.update([(nums[i],nums[j]), (nums[j],nums[i]), (nums[k],nums[j]), (nums[j],nums[k]), (nums[k],nums[i]), (nums[i],nums[k])])

		return result



O(n^2)

###################################################

class Solution:
	def threeSum(self, nums: List[int]) -> List[List[int]]:
		sorted_nums = sorted(nums)
		
		n = len(sorted_nums)
		result = []
		idx = 0
		while idx < n-2:
		
			curr_target = sorted_nums[idx]
			left = idx + 1
			right = n-1

			while left < right:
				sum = sorted_nums[left] + sorted_nums[right] 
				if sum  == -curr_target:
					result.append([sorted_nums[idx], sorted_nums[left], sorted_nums[right]])
					left_val = sorted_nums[left]
					right_val = sorted_nums[right] 
					while left < right and sorted_nums[left] == left_val:
						left += 1

					while right > left and sorted_nums[right] == right_val:
						right -= 1
			
				elif sum > -curr_target:
					right -= 1
				else:
					left += 1


			while idx < n-2 and sorted_nums[idx] == curr_target:
				idx += 1							
			

		return result
 




#####################################3

























		
		
		